modules/core/src/main/java/org/apache/ignite/internal/processors/cache/transactions/IgniteTxManager.java

```diff
@@ -939,11 +935,13 @@ public GridNearTxLocal newTx(
                 });
 
         for (IgniteInternalTx tx : activeTransactions()) {
-            if (tx.dht() && !tx.local() && tx.originatingNodeId().equals(node.id())) {
+            if (tx.state() == PREPARED /* recovery may be needed */
```

Comment:
This doesn't look correct to me. The transaction can be already in COMMITTING state, resulting in the exchange future finished before counters are applied due to race, and broken invariant HWM >= LWM as a consequence.

Reply:
@alex-plekhanov had a similar concern, but about the PREPARING state.
Fixed to any state, because we should wait for the finishing of everything related to the failed node.